### PR TITLE
[14.0][ADD] report_xlsx_pdf: print directly to PDF from xlsx

### PR DESCRIPTION
**Pull request by TheMule71**
Extend report xlsx to export directly to pdf.

### CODE REVIEW

**Review by andreampiovesana — APPROVED** [review]
ok

**Comment by andreampiovesana** [comment]
add in manifest
   "external_dependencies": {
       "python": ["xlsxwriter"],
       "deb": ["libreoffice"],
   },

**Comment by scigghia** [comment]
why this error in runboat?

++ oca_list_external_dependencies deb
+ DEBIAN_FRONTEND=noninteractive
+ apt-get install -qq --no-install-recommends default-jre-headless graphviz libreoffice
E: Failed to fetch http://security.ubuntu.com/ubuntu/pool/main/p/python3.8/libpython3.8_3.8.10-0ubuntu1~20.04.4_amd64.deb  404  Not Found [IP: 185.125.190.36 80]
E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?

however for me it is ok

**Review by scigghia — APPROVED** [review]

File: report_xlsx_pdf/readme/INSTALL.rst [hunk 1]
```diff
@@ -0,0 +1,7 @@
+Make sure you have ``xlsxwriter`` Python module installed::
+
+$ pip3 install xlsxwriter
+
+For testing it is also necessary ``xlrd`` Python module installed::
+
+$ pip3 install xlrd
```
Comment (primes2h) [inline report_xlsx_pdf/readme/INSTALL.rst:10]: I would add libreoffice installed as a requirement as well (only on Linux systems for now).

File: report_xlsx_pdf/models/ir_report.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+from odoo.tools.misc import find_in_path
+
+logger = logging.getLogger(__name__)
+
+
+class ReportAction(models.Model):
+    _inherit = "ir.actions.report"
+
+    report_xlsx_to_pdf = fields.Boolean(
+        string="Report xlsx to pdf",
```
Comment (primes2h) [inline report_xlsx_pdf/models/ir_report.py]:
```suggestion
        string="Export to pdf format",
```

File: report_xlsx_pdf/__manifest__.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
+# Copyright 2022 Openindustry.it (<https://openindustry.it>)
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+{
+    "name": "Extend report xlsx to export directly to pdf",
+    "summary": "Extend report xlsx to export directly to pdf",
```
Comment (primes2h) [inline report_xlsx_pdf/__manifest__.py]:
```suggestion
    "name": "Export xlsx report to pdf",
    "summary": "Extend report_xlsx to export directly to pdf",
```

File: report_xlsx_pdf/readme/DESCRIPTION.rst [hunk 4]
```diff
@@ -0,0 +1 @@
+This module extend xlsx report to convert in pdf using libreoffice
```
Comment (primes2h) [inline report_xlsx_pdf/readme/DESCRIPTION.rst]:
```suggestion
This module extend ``report_xlsx`` to convert xlsx to pdf using libreoffice.
```

**Review by primes2h — CHANGES_REQUESTED** [review]

**Review by marcelofrare — APPROVED** [review]

**Comment by TheMule71** [comment]
> why this error in runboat?
> 
> ++ oca_list_external_dependencies deb
> 
> * DEBIAN_FRONTEND=noninteractive
> * apt-get install -qq --no-install-recommends default-jre-headless graphviz libreoffice
>   E: Failed to fetch http://security.ubuntu.com/ubuntu/pool/main/p/python3.8/libpython3.8_3.8.10-0ubuntu1~20.04.4_amd64.deb  404  Not Found [IP: 185.125.190.36 80]
>   E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?
> 
> however for me it is ok

I'm at loss here.  It's looking for the wrong file, the correct link is:

`http://security.ubuntu.com/ubuntu/pool/main/p/python3.8/libpython3.8_3.8.10-0ubuntu1~20.04.5_amd64.deb`

not

`http://security.ubuntu.com/ubuntu/pool/main/p/python3.8/libpython3.8_3.8.10-0ubuntu1~20.04.4_amd64.deb`

but I'm not sure how to fix it.

@sbidoul (sorry to bother you)
Maybe do an `apt-get update`
before this `apt-get install` here?
https://github.com/OCA/oca-ci/blob/cffeb8853b2833c5a47da3200b4e3025b47de043/bin/oca_install_addons#L28

I might be completely off track, tho. I've never looked at those scripts before.

**Comment by primes2h** [comment]
@TheMule71 
Please mark PR as ready for review.

**Comment by TheMule71** [comment]
> @TheMule71 Please mark PR as ready for review.

OK, but tests are still failing.

**Comment by sbidoul** [comment]
> Maybe do an apt-get update before this apt-get install here?

@TheMule71 yes that's what we need to do. There is an open issue for that in oca/oca-ci (https://github.com/OCA/oca-ci/issues/26). Would you like to do a PR for that ?

**Comment by etobella** [comment]
Now it is green and ready to review :smile:

**Comment by TheMule71** [comment]
> Now it is green and ready to review smile

Thanks!



> > Maybe do an apt-get update before this apt-get install here?
> 
> @TheMule71 yes that's what we need to do. There is an open issue for that in oca/oca-ci ([OCA/oca-ci#26](https://github.com/OCA/oca-ci/issues/26)). Would you like to do a PR for that ?

Looks like @etobella beat me to that. :)

**Review by francescapenso — APPROVED** [review]

**Review by primes2h — APPROVED** [review]
LGTM

Thanks!

**Comment by andreampiovesana** [comment]
merge?

File: report_xlsx_pdf/readme/USAGE.rst [hunk 5]
```diff
@@ -0,0 +1,31 @@
+An example of XLSX report for partners on a module called `module_name`:
```
Comment (etobella) [inline report_xlsx_pdf/readme/USAGE.rst:1]: This usage comes from report_xlsx it is probably not needed, isn't it?

The real usage is showing how the flag is used, Can you change it?

File: report_xlsx_pdf/readme/INSTALL.rst [hunk 6]
```diff
@@ -0,0 +1,10 @@
+Make sure you have ``xlsxwriter`` Python module installed::
```
Comment (etobella) [inline report_xlsx_pdf/readme/INSTALL.rst:1]: This comes from report_xlsx, so it is probably not needed here. The only thing needed is libreoffice in this case, as xlsxwriter comes from report_xlsx

File: report_xlsx_pdf/tests/test_report.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+class TestReport(common.TransactionCase):
+    def setUp(self):
+        super(TestReport, self).setUp()
+        report_object = self.env["ir.actions.report"]
+        self.xlsx_report = self.env["report.report_xlsx.abstract"].with_context(
+            active_model="res.partner"
+        )
+        self.report_name = "report_xlsx.partner_xlsx"
+        self.report_xlsx_to_pdf = True
```
Comment (etobella) [inline report_xlsx_pdf/tests/test_report.py:24]: I don't understand this. It is never used... You never tested that is is transformed. This test is for report_xlsx standard. Can you test the code from this module?

**Review by etobella — CHANGES_REQUESTED** [review]
Readme is not complete.

Tests are for report_xlsx, can you test the transformation?

File: report_xlsx_pdf/tests/test_report.py (hunk 7 above)
Comment (andreampiovesana) [reply report_xlsx_pdf/tests/test_report.py:24]: self.report_xlsx_to_pdf = True
is the new parameter for this app

**Comment by andreampiovesana** [comment]
for new merge of PR #628
you shoud change
image

**Comment by andreampiovesana** [comment]
merge?

**Comment by francesco-ooops** [comment]
> merge?

@andreampiovesana have @etobella 's remarks been addressed?

**Comment by andreampiovesana** [comment]
self.report_xlsx_to_pdf = True

is the specific parameter

right?


Il giorno sab 27 ago 2022 alle ore 22:13 Enric Tobella <
***@***.***> ha scritto:

**Comment by andreampiovesana** [comment]
@OCA/reporting-engine-maintainers we use this app in production and this fix many situations where print in qweb-pdf take too much time to export, specially with report with hundreds of pages and we use the same technology of py3o. 
Please can you merge?

**Comment by HviorForgeFlow** [comment]
ping @etobella

File: report_xlsx_pdf/tests/test_report.py (hunk 7 above)
Comment (etobella) [reply report_xlsx_pdf/tests/test_report.py:24]: But you are editing self, in this case, the TestClass, it does not make sense. You should do `self.xlsx_report.report_xlsx_to_pdf = True`

**Comment by etobella** [comment]
Reopening, as the necessary change is minimal. Can you review it @TheMule71 @andreampiovesana

File: report_xlsx_pdf/tests/test_report.py (hunk 7 above)
Comment (HviorForgeFlow) [reply report_xlsx_pdf/tests/test_report.py:24]: @andreampiovesana can you attend this comment? Thanks!

File: report_xlsx_pdf/__manifest__.py [hunk 8]
```diff
@@ -0,0 +1,20 @@
+# Copyright 2022 Openindustry.it (<https://openindustry.it>)
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+{
+    "name": "Export xlsx report to pdf",
+    "summary": "Extend report_xlsx to export directly to pdf",
+    "author": "Openindustry.it," "Odoo Community Association (OCA)",
```
Comment (HviorForgeFlow) [inline report_xlsx_pdf/__manifest__.py:6]:
```suggestion
    "author": "Openindustry.it, Odoo Community Association (OCA)",
```

**Review by HviorForgeFlow — CHANGES_REQUESTED** [review]
Minor changes required, after that, it can be merged :)

**Pull request closed without merging**